File: bindings/typescript/src/identity/authoring.ts
import { createRawKeyEd25519IdentityClient, type ValidatedIdentity } from "../identity.js";
import type { VerificationRelationship } from "./adapters.js";
import { loadPackagedWorkflowEngine } from "../verifier/wasm.js";

declare const preparedIdentityMessageBrand: unique symbol;
export interface PreparedIdentityMessage {
  readonly [preparedIdentityMessageBrand]: true;
  readonly identity: ValidatedIdentity;
  readonly relationshipId: string;
  readonly message: Uint8Array;
  readonly signingPreimage: Uint8Array;
}

interface AuthoringEngine {
  createRawKeyPublicIdentityV2(suiteId: string, publicKey: Uint8Array): Uint8Array;
  encodeIdentityDescriptorV1(value: unknown): Uint8Array;
  identityDescriptorSigningPreimageV1(packet: Uint8Array, relationshipId: string, message: Uint8Array): Uint8Array;
  identityMessageSigningPreimageV2(packet: Uint8Array, message: Uint8Array): Uint8Array;
}

export async function createRawKeyEd25519Identity(publicKey: Uint8Array): Promise<ValidatedIdentity> {
  if (!(publicKey instanceof Uint8Array) || publicKey.length !== 32) throw new TypeError("Ed25519 public key must contain exactly 32 bytes");
  const engine = await loadPackagedWorkflowEngine() as unknown as AuthoringEngine;
  const packet = new Uint8Array(engine.createRawKeyPublicIdentityV2("ed25519-v1", publicKey.slice()));
  const client = await createRawKeyEd25519IdentityClient();
  try {
    const decoded = client.decode(packet); if (decoded.kind !== "ok") throw new TypeError(decoded.issue.code);
    const resolved = await client.resolve(decoded.value); if (resolved.kind !== "ok") throw new TypeError(resolved.issue.code);
    const validated = await client.validate(resolved.value); if (validated.kind !== "ok") throw new TypeError(validated.issue.code);
    return validated.value;
  } finally { await client.close(); }
}

export async function encodeIdentity(input: Readonly<{ methodId: string; identityId: string; methodMaterial?: Uint8Array; relationships: readonly VerificationRelationship[] }>): Promise<Uint8Array> {
  const engine = await loadPackagedWorkflowEngine() as unknown as AuthoringEngine;
  return new Uint8Array(engine.encodeIdentityDescriptorV1({
    methodId: input.methodId,
    identityId: input.identityId,
    methodMaterial: input.methodMaterial?.slice() ?? new Uint8Array(),
    relationships: input.relationships.map((relationship) => ({
      relationshipId: relationship.id,
      purpose: relationship.purpose,
      suiteId: relationship.suiteId,
      verificationMaterial: relationship.verificationMaterial.map((material) => ({ materialId: material.id, bytes: material.bytes.slice() })),
    })),
  }));
}

export async function prepareIdentityMessage(input: Readonly<{ identity: ValidatedIdentity; relationshipId?: string; message: Uint8Array }>): Promise<PreparedIdentityMessage> {
  const relationshipId = input.relationshipId ?? "default-signing";
  if (!input.identity.relationships.includes(relationshipId)) throw new TypeError("identity relationship is unavailable");
  const engine = await loadPackagedWorkflowEngine() as unknown as AuthoringEngine;
  const packet = input.identity.toBytes();
  let signingPreimage: Uint8Array;
  try { signingPreimage = new Uint8Array(engine.identityDescriptorSigningPreimageV1(packet, relationshipId, input.message.slice())); }
  catch { signingPreimage = new Uint8Array(engine.identityMessageSigningPreimageV2(packet, input.message.slice())); }
  return Object.freeze({ identity: input.identity, relationshipId, message: input.message.slice(), signingPreimage }) as PreparedIdentityMessage;
}

File: bindings/python/python/auths/identity/authoring.py
from __future__ import annotations

from typing import Sequence as _Sequence

from .._native import (
    compact_identity_descriptor_v1 as _compact_descriptor,
    decode_identity_descriptor_v1 as _decode_descriptor,
    encode_identity_descriptor_v1 as _encode_descriptor,
    identity_descriptor_signing_preimage_v1 as _signing_preimage,
    raw_key_identity_v2 as _raw_key_identity,
)
from . import ValidatedIdentity, _validated
from .adapters import DecodedIdentityRecord, ResolutionEvidence, ResolvedIdentityRecord, VerificationMaterial, VerificationRelationship

_TOKEN = object()


class PreparedIdentityMessage:
    __slots__ = ("identity", "relationship_id", "message", "signing_preimage")

    def __init__(self, token: object, identity: ValidatedIdentity, relationship_id: str, message: bytes, preimage: bytes) -> None:
        if token is not _TOKEN:
            raise TypeError("PreparedIdentityMessage is sealed")
        self.identity = identity
        self.relationship_id = relationship_id
        self.message = bytes(message)
        self.signing_preimage = bytes(preimage)


def create_raw_key_ed25519_identity(public_key: bytes, /) -> ValidatedIdentity:
    key = bytes(public_key)
    if len(key) != 32:
        raise ValueError("Ed25519 public key must contain exactly 32 bytes")
    packet = bytes(_compact_descriptor(_raw_key_identity("ed25519-v1", key)))
    projection = _decode_descriptor(packet)
    relationships = tuple(
        VerificationRelationship(
            relationship_id,
            purpose,
            suite_id,
            tuple(
                VerificationMaterial(material_id, bytes(material))
                for material_id, material in materials
            ),
        )
        for relationship_id, purpose, suite_id, materials in projection.relationships
    )
    decoded = DecodedIdentityRecord(
        projection.method_id,
        projection.identity_id,
        bytes(projection.method_material),
        relationships,
    )
    evidence = ResolutionEvidence("authoring", 0, 2**63 - 1, ("self-contained",))
    return _validated(packet, ResolvedIdentityRecord(decoded.method_id, decoded.identity_id, decoded.method_material, decoded.relationships, evidence))


def encode_identity(*, method_id: str, identity_id: str, relationships: _Sequence[VerificationRelationship], method_material: bytes = b"") -> bytes:
    values = tuple(relationships)
    if not 1 <= len(values) <= 16:
        raise ValueError("identity requires 1..16 relationships")
    native = [
        (value.relationship_id, value.purpose, value.suite_id, [(material.material_id, bytes(material.bytes)) for material in value.verification_material])
        for value in values
    ]
    return bytes(_encode_descriptor(method_id, identity_id, bytes(method_material), native))


def prepare_identity_message(identity: ValidatedIdentity, /, *, message: bytes, relationship_id: str = "default-signing") -> PreparedIdentityMessage:
    if relationship_id not in identity.relationships:
        raise ValueError("relationship is not present on the validated identity")
    body = bytes(message)
    if len(body) > 65_536:
        raise ValueError("identity message exceeds 65536 bytes")
    return PreparedIdentityMessage(_TOKEN, identity, relationship_id, body, _signing_preimage(identity.to_bytes(), relationship_id, body))


__all__ = [
    "PreparedIdentityMessage", "create_raw_key_ed25519_identity",
    "encode_identity", "prepare_identity_message",
]
